### PR TITLE
access log formatter: use new formatter context rather than multiple parameters (3/3)

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: access log formatter: use new formatter context rather than multiple parameters (3/3)
Additional Description:

Continuous work of #30757. Waiting #30757 to be landed first.

The PR update the log `Instance` interface to be a specialization of `InstanceBase` with context type HttpFormatterContext.
And by this way, the `log()` method of Filter is updated to take a HttpFormatterContext reference as parameter.

This PR also do similar update to the `AccessLogInstanceFactory` and make it a specialization of AccessLogInstanceFactoryBase.

By this way, log instances of different modules will share the same code base.

Note, to ensure the backwards compatibility, the extension category of log instance should be kept as envoy.access_loggers.

This is the last PR of this big refactoring. After landed this PR, we will finally completely migrate our formatter/logger to use the context based interface.

This will bring two obvious benefits:
1. it's easier to make the common part of these code be a template. All formatter-based loggers could be shared by different modules like dubbo proxy, generic proxy or even redis/mysql if necessary (some minor update is still necessary, but it's simple and minor).
2. it's easier to extend the logger/formatter to add new feature like request trailer support. We needn't to change almost every log()/format() call to add a feature that only used in part of positions anymore. (think of the access log type 😥 ).

Risk Level: low.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
Related to https://github.com/envoyproxy/envoy/issues/13085

File: source/common/http/filter_manager.h [hunk 1]
```diff
@@ -664,22 +664,13 @@ class FilterManager : public ScopeTrackedObject,
...
-      response_headers = filter_manager_callbacks_.responseHeaders().ptr();
-    }
-    ResponseTrailerMap* response_trailers = nullptr;
-    if (filter_manager_callbacks_.responseTrailers()) {
-      response_trailers = filter_manager_callbacks_.responseTrailers().ptr();
-    }
+    Formatter::HttpFormatterContext log_context{filter_manager_callbacks_.requestHeaders().ptr(),
+                                                filter_manager_callbacks_.responseHeaders().ptr(),
+                                                filter_manager_callbacks_.responseTrailers().ptr()};
+    log_context.setAccessLogType(access_log_type);
```
Comment (RyanTheOptimist) [inline source/common/http/filter_manager.h:670]: Can we add a 4-arg constructor so we can eliminate all these calls to `setAccessLogType()` and instead just pass it in to the constructor.

**Review by RyanTheOptimist — COMMENTED** [review]
LGTM, but one suggestion.

File: source/common/http/filter_manager.h (hunk 1 above)
Comment (wbpcode) [reply source/common/http/filter_manager.h:670]: In fact, we have a 5-arg constructor, let's use it directly.

Comment (wbpcode) [reply source/common/http/filter_manager.h]: done.

**Comment by wbpcode** [comment]
weird coverage result. the result tell wasm coverage rate is lower than the limit but from my local test result, it should be covered all.

**Review by RyanTheOptimist — APPROVED** [review]
Looks great. Thanks for doing this.

**Pull request merged**